api/client/alpn_conn_upgrade.go

```diff
@@ -246,14 +254,28 @@ func upgradeConnThroughWebAPI(conn net.Conn, api url.URL, upgradeType string) (n
 			return nil, trace.NotImplemented(
 				"connection upgrade call to %q with upgrade type %v failed with status code %v. Please upgrade the server and try again.",
 				constants.WebAPIConnUpgrade,
-				upgradeType,
+				alpnUpgradeType,
 				resp.StatusCode,
 			)
 		}
 		return nil, trace.BadParameter("failed to switch Protocols %v", resp.StatusCode)
 	}
 
-	if upgradeType == constants.WebAPIConnUpgradeTypeALPNPing {
+	// Handle WebSocket.
+	if resp.Header.Get(constants.WebAPIConnUpgradeHeader) == constants.WebAPIConnUpgradeTypeWebSocket {
+		if err := checkWebSocketUpgradeResponse(resp, alpnUpgradeType, challengeKey); err != nil {
+			defer conn.Close()
+			return nil, trace.Wrap(err)
+		}
+
+		logrus.Debugf("Performing ALPN WebSocket connection upgrade for %v.", api.Host)
+		return newWebSocketALPNClientConn(conn), nil
+	}
+
+	// Handle "legacy".
+	// TODO(greedy52) DELETE in 17.0.
+	logrus.Debugf("Performing ALPN legacy connection upgrade for %v.", api.Host)
```

Comment:
As of RFD 0154 (https://github.com/gravitational/teleport/blob/master/rfd/0154-logging-guidelines.md#withfields), we prefer using `WithFields` rather than sprintf style log messages.